crates/runtime/src/dataconnector/sharepoint.rs

```diff
@@ -0,0 +1,243 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use crate::component::dataset::Dataset;
+use async_trait::async_trait;
+use data_components::sharepoint::client::SharepointClient;
+use datafusion::datasource::TableProvider;
+use graph_rs_sdk::{
+    error::AuthorizationFailure,
+    identity::{AuthorizationCodeCredential, ConfidentialClientApplication},
+    GraphClient,
+};
+use snafu::{ResultExt, Snafu};
+use std::any::Any;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use url::Url;
+
+use super::{
+    DataConnector, DataConnectorFactory, DataConnectorResult, ParameterSpec, Parameters,
+    UnableToGetReadProviderSnafu,
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Missing required parameter: {parameter}"))]
+    MissingParameter { parameter: String },
+
+    #[snafu(display("Missing redirect url parameter: {url}"))]
+    InvalidRedirectUrlError { url: String },
+
+    #[snafu(display("Could not construct authorization url for sharepoint OAuth2: {source}"))]
+    AuthorizationUrlError {
+        source: graph_rs_sdk::error::AuthorizationFailure,
+    },
+
+    /// Error when user hasn't clicked authorisation URL yet. If this error occurs, can still be valid on retry.
+    #[snafu(display("No authorisation code yet available to access sharepoint data"))]
+    AuthorizationCodeError {},
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct Sharepoint {
+    client: Arc<GraphClient>,
+}
+
+/// Create a [`GraphClient`] from a client authorization code (i.e. OAuth 2.0 authorization code flow).
+/// `<https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow>`
+#[must_use]
+pub fn auth_code_grant_secret(
+    authorization_code: &str,
+    client_id: &str,
+    client_secret: &str,
+    scope: Vec<String>,
+    redirect_uri: Url,
+) -> GraphClient {
+    GraphClient::from(
+        &ConfidentialClientApplication::builder(client_id)
+            .with_auth_code(authorization_code)
+            .with_client_secret(client_secret)
+            .with_scope(scope)
+            .with_redirect_uri(redirect_uri)
+            .build(),
+    )
+}
+
+/// [`Url`] for users to sign in to the application. Returns state + authorization URL.
+pub fn authorization_sign_in_url(
+    client_id: &str,
+    state: &str,
+    tenant_id: &str,
+    redirect_uri: Url,
+    scope: Vec<String>,
+) -> std::result::Result<Url, AuthorizationFailure> {
+    AuthorizationCodeCredential::authorization_url_builder(client_id)
+        .with_redirect_uri(redirect_uri)
+        .with_scope(scope)
+        .with_tenant(tenant_id)
+        .with_state(state)
+        .url()
+}
+
+static NEEDED_SCOPE: [&str; 2] = ["User.Read", "Files.Read"];
+static REDIRECT_URL: &str = "http://localhost:8090/redirect/sharepoint";
```

Comment:
The port can be changed so 8090 should not be hardcoded.

Reply:
Not needed anymore, hopefully